Add a KEP for Node-Scoped Daemonset

haiyanmeng opened the pull request:
This KEP describes a mechanism restricting a DaemonSet pod to only managing the resources that reside on the same node as the DaemonSet pod.

Tracking issue: the 4.c part of https://github.com/kubernetes/kubernetes/issues/62747

haiyanmeng commented:
/cc @tallclair

janetkuo commented:
/ok-to-test

haiyanmeng commented:
/cc @janetkuo

janetkuo commented:
/sig apps

haiyanmeng commented on keps/sig-auth/20190408-node-scoped-daemonset.md:
```diff
@@ -0,0 +1,239 @@
+          to run multiple DaemonSets which have different scope requirements;
+        * this can be used to set the scope of non-daemonset pods, which might
+          be confusing.
+* Option 2: specify the node-scoped info in DaemonSet either in its
```
@janetkuo , the node-scoped info can be specified either in ServiceAccount (option 1) or DaemonSet (option 2).
Currently, we favor option 1.

tallclair commented on keps/sig-auth/20190408-node-scoped-daemonset.md:
```diff
@@ -0,0 +1,239 @@
+  - "@haiyanmeng"
+owning-sig: sig-auth
+participating-sigs:
+  - sig-auth
```
nit: I think the owning-sig is implied here. I'd add sig-node though.

haiyanmeng replied:
Fixed it! Thanks!

tallclair commented on keps/sig-auth/20190408-node-scoped-daemonset.md:
```diff
@@ -0,0 +1,239 @@
+    * Cons:
+        * this may not work for the cases where a single service account is used
+          to run multiple DaemonSets which have different scope requirements;
+        * this can be used to set the scope of non-daemonset pods, which might
```
IIUC, option 2 also has this disadvantage. From the kubelet's perspective, a daemonset pod is just another pod - it doesn't matter where it came from.

haiyanmeng replied:
True. 
The difference is that:
using option 1, the non-daemonset pods will be affected as soon as it runs with a service account including node-scoped info;
using option 2, the non-daemonset pods need to set the node-scoped info explicitly to get affected.

I will update the doc to reflect the difference. Thanks!

haiyanmeng replied:
Fixed it! Thanks!

tallclair commented on keps/sig-auth/20190408-node-scoped-daemonset.md:
```diff
@@ -0,0 +1,239 @@
+* At this point, NodeAuthorizer has all the necessary information to make its
+  authorization decision. The graph below shows the decision making process.
+
+![The decision-making process of NodeAuthorizer](node-scoped-daemonset.png)
```
I think it would be worth enumerating the differences between the existing NodeAuthorizer logic, and the new logic. As written, the graph makes me think node-scoped daemonsets should get a new authorizer plugin (but I think they're actually fairly similar).

haiyanmeng replied:
Since the latest version of the KEP has changed so much, mark this one as obsolete. Thanks!

tallclair commented on keps/sig-auth/20190408-node-scoped-daemonset.md:
```diff
@@ -0,0 +1,239 @@
+DaemonSet.
+
+With the introduction of
+[BoundServiceAccountToken](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/auth/bound-service-account-tokens.md)
```
BoundServiceAccountTokens are set up by the Kubelet, and the credential is obtained through a token request. Is there already logic in the NodeRestriction controller to prevent the Kubelet identity from requesting a token bound to a pod on another node?

haiyanmeng replied:
Yes. NodeRestriction already covers this:
https://github.com/kubernetes/kubernetes/blob/3ec638d3cd75794dc7b034389b5065f7139b8570/plugin/pkg/admission/noderestriction/admission.go#L142

tallclair commented on keps/sig-auth/20190408-node-scoped-daemonset.md:
```diff
@@ -0,0 +1,239 @@
+  node-scoped info into the Extra field of user.DefaultInfo;
+* NodeRestriction gets the pod info and the node-scoped info from the Extra
+  field of user.DefaultInfo through GetExtra; using the pod info,
+  NodeRestriction can gets the node information of P1 by checking
```
I thought it could just fetch it directly from the graph?

haiyanmeng replied:
Do you mean the resource graph used in NodeAuthorizer?

Since NodeRestriction can inspect the request objects, there is no need to check the resource graph.

tallclair commented on keps/sig-auth/20190408-node-scoped-daemonset.md:
```diff
@@ -0,0 +1,239 @@
+// isKubelet will be set to true, and nodeName will be set to the node name; if the
+// request sender is a node-scoped DaemsonSet pod, isNodeScopedDaemonSet will be
+// set to true, and nodeName will be set to the node name.
+NodeIdentity(user.Info, NodeInfo) (nodeName string, isKubelet bool, isNodeScopedDaemonSet bool)
```
nit: I'd like to avoid the extra boolean. Maybe use an enum instead?

haiyanmeng replied:
Fixed it! Thanks!

tallclair commented on keps/sig-auth/20190408-node-scoped-daemonset.md:
```diff
@@ -0,0 +1,239 @@
+
+v1.15:
+
+* implement the proposal, and extend the existing unit and integration tests.
```
implement the proposal in alpha, right? This should have a dedicated feature gate, and will probably need an API review.

derekwaynecarr replied:
Agree with dedicated feature gate

haiyanmeng replied:
Thanks for pointing out. Fixed it!

tallclair commented on keps/sig-auth/20190408-node-scoped-daemonset.md:
```diff
@@ -0,0 +1,239 @@
+
+* Option 1: specify the node-scoped info as an annotation (key:
+  `scope.k8s.io/node-scoped`; value: `true` or `false` (default)) or a new
+  boolean field in
```
Probably a boolean field. I think functional annotations are no longer allowed.

haiyanmeng replied:
Just to be clear, are functional annotations no longer allowed in all the API objects? Or just in ServiceAccount?

tallclair replied:
In API objects, at least for in-tree functionality.

haiyanmeng replied:
I see. Thanks!

tallclair commented:
Nodes from [2019-04-17 sig-auth](https://docs.google.com/document/d/1woLGRoONE3EBVx-wTb4pvp4CI7tmLZ6lS26VTbosLKM/edit#heading=h.5z900oolyun5):

- Consider separating the authorization & admission piece. It looks like the admission piece doesn't need the graph, and is more node specific. Admission could be handled with policy controls (e.g. gatekeeper)
- Intersecting permissions has been previously discussed, and might make sense here - would intersecting service account & kubelet permissions be sufficient?
- Some questions about whether this is sufficiently generic, and how it applies to non-DaemonSet workloads.
  - _Note(tallclair): I don't see the use case for non-daemonset workloads. IMO this only makes sense when the workload is node-aware_

Anything else I missed?

haiyanmeng commented:
/cc @deads2k @smarterclayton

haiyanmeng commented:
Based on the feedback from last week's sig-auth meeting, I updated the KEP. The main changes are:
1) added a `Use Cases` section and `Auternatives` section;
2) updated the `Proposal` section. In short, the new solution allows the DaemonSet owners to specify a list of resource types whose access permissions should be node-scoped; during the authorization step, the requests from node-scoped daemonset pods will go through both NodeAuthorizer and RBAC, and the request must be allowed by the NodeAuthorizer first before RBAC authorizes the request. If NodeAuthorizer does not explicitly allow the request, the request will be denied directly.

Please let me know if you have any suggestions.

enj commented on keps/sig-auth/20190408-node-scoped-daemonset.md:
```diff
@@ -0,0 +1,378 @@
+
+Node-scoped info can be defined in two places:
+
+* Option 1: in
```
I am still trying to grasp all of the pieces here, but adding a new field to SA seems wrong to me.  Could this be handled via the token request API?

mikedanese commented on keps/sig-auth/20190408-node-scoped-daemonset.md:
```diff
@@ -0,0 +1,378 @@
+
+### The Origin of Node-Scoped Resource Spec
+
+Node-scoped info can be defined in two places:
```
Both these options feel like authorization bleeding out of authorization. We unfortunately do this in various places already but it hurts our ability to build authorization features, e.g. can-i, explain, delegated authz.

haiyanmeng commented:
As discussed in the sig-auth meeting, the short-term plan to achieve node-scoped daemonsets is to run daemonsets with kubelet credentials, while we work on a long term plan, which will avoid tangling authorizers together and may conditionalize authorization.

cheyang commented:
@haiyanmeng May I ask if we'd like to use [Node-Scoped Daemonset](https://github.com/kubernetes/enhancements/pull/944#top). What's the recommended way?  Still with kubelet credentials?
